### ISSUE TITLE
When browser doesn't have the redux devtools extension the app throws an undefined error

### ISSUE THREAD

**Issue by Arrow7000** [zalmoxisus/redux-devtools-extension#320]
My React/Redux app worked fine on my computer but when I sent it to someone else the app would fail to load altogether. 

After some debugging it turns out that `window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()` was returning undefined, which meant that `.apply()` was being called on `undefined`, thereby causing the error. 

I therefore propose to change this line in the documentation to:

```
window.__REDUX_DEVTOOLS_EXTENSION__ ? window.__REDUX_DEVTOOLS_EXTENSION__() : f => f
```

This will make sure that if the redux extension is not present, the store enhancer added will at least be a function, as opposed to `undefined`.

I will submit a pull request immediately.

**Comment by zalmoxisus** [comment]
That's because you're including it inside the compose. In this case you should use `window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose`. See [Advanced store setup](https://github.com/zalmoxisus/redux-devtools-extension#12-advanced-store-setup)

**Comment by Arrow7000** [comment]
Hm ok I see. Does the `__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ` include additional functionality? Because if it doesn't it looks like my solution is a little bit cleaner. Would obviate the need for an additional global variable.

I haven't tested this on a basic store, but if it *does* work on a basic store then maybe it could be nicer to only have a single line of code that will add the functionality for both? 

What do you think?

**Comment by zalmoxisus** [comment]
Yes, it does. Because the enhancer is last in the compose, it is not aware of other middleware and enhancers, thus remotely dispatched actions wouldn't work as expected and `lock` button will not have effect on enhancers.

See [the blog post](https://medium.com/@zalmoxis/improve-your-development-workflow-with-redux-devtools-extension-f0379227ff83) for more details why we introduced this.

**Comment by zalmoxisus** [comment]
`__REDUX_DEVTOOLS_EXTENSION_COMPOSE__` can be used instead of `__REDUX_DEVTOOLS_EXTENSION__`, but it's confusing and we might not support it in future, see https://github.com/zalmoxisus/redux-devtools-extension/issues/310.

**Comment by zalmoxisus** [comment]
To clarify, you should use one of them for the specific case, not both. `__REDUX_DEVTOOLS_EXTENSION__` is used only when it's the only enhancer, not to be included in the compose.

**Comment by Arrow7000** [comment]
Ok thank you. I got started with the basic store setup and didn't look at the advanced setup when I started including enhancers.

Thanks for explaining things so clearly!

**Comment by zalmoxisus** [comment]
You're welcome. We could add a note in the README for the basic store and information about that exception in troubleshooting. If you change your pr with that, it would be much appreciated.

**Comment by Arrow7000** [comment]
Ok will do

**Comment by hackhat** [comment]
I wonder why redux is not checking for `undefined`? Is assuming that all args are a function, but sometimes we want to pass `undefined`.

**Comment by hugoh1995** [comment]
I'm also getting the same issue. Can't find a fix.

**Comment by danielecr** [comment]
@Shadowman4205 that solution does not work for me too, instead I tryed with


> enhancer = applyMiddleware(settingsMiddleware,  epicMiddleware, router);

and this works, also `applyMiddleware()` accept an array as parameter, so I think that applyMiddleware(...middleware) is meant as:

> enhancer = compose( applyMiddleware([settingsMiddleware,  epicMiddleware, router]) )

but I am still confused

**Comment by AndrewCraswell** [comment]
@Shadowman4205 I was having the same issue until I changed it slightly to match the Advanced Setup guide:

```
  const composeEnhancers =
    typeof window === 'object' &&
      (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?
      (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) : compose;
```

Here I'm passing an empty `{}` object to the `__REDUX_DEVTOOLS_EXTENSION_COMPOSE__()`. ONly then it works, otherwise throws the same error you're receiving. This feels counterintuitive to me.

**Comment by sht5** [comment]
see solution through package here(worked for me):
https://stackoverflow.com/questions/49190781/redux-app-not-working-if-redux-devtools-extension-is-not-installed

**Comment by qodesmith** [comment]
I needed to go from this:
```
const composeEnhancers = !__PROD__ ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose
```

To this:
```
const composeEnhancers = !__PROD__ ? (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose) : compose
```

**Comment by KrishieldKyle** [comment]
i fixed mine, now my app can now run with or without reduxDevtool installed

i just changed 

`window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()` 

to

`window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__() || compose`

**Comment by mmunier44** [comment]
Bottom one worked for me thanks so much

**Comment by mhs1** [comment]
@KrishieldKyle: thx!

**Comment by kekexunxun** [comment]
@Arrow7000  thx
